File: tests/fixtures/selenium-chaos-suite-csharp/Utilities/WaitUtils.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.UI;

namespace SeleniumChaosSuite.Utilities;

public static class WaitUtils
{
    internal static Task Sleep(int ms) => Task.Delay(ms);

    public static IWebElement WaitVisible(IWebDriver driver, By locator, int timeoutMs = 20000)
    {
        var wait = new WebDriverWait(driver, TimeSpan.FromMilliseconds(timeoutMs));
        return wait.Until(drv =>
        {
            var el = drv.FindElement(locator);
            return el.Displayed ? el : null;
        })!;
    }

    public static bool WaitEnabled(IWebElement element, int timeoutMs = 10000)
    {
        var started = DateTime.UtcNow;
        while ((DateTime.UtcNow - started).TotalMilliseconds < timeoutMs)
        {
            try
            {
                if (element.Enabled) return true;
            }
            catch
            {
                // Ignore while polling.
            }

            Thread.Sleep(250);
        }

        return false;
    }

    public static void RetryClick(IWebDriver driver, By locator, int tries = 5, int delayMs = 300)
    {
        for (var i = 0; i < tries; i++)
        {
            try
            {
                var wait = new WebDriverWait(driver, TimeSpan.FromMilliseconds(delayMs));
                var el = wait.Until(d => d.FindElement(locator));
                el.Click();
                return;
            }
            catch when (i < tries - 1)
            {
                Thread.Sleep(delayMs);
            }
        }
    }

    public static bool ScrollUntilTextAppears(IWebDriver driver, string text, int rounds = 12)
    {
        for (var i = 0; i < rounds; i++)
        {
            var bodyText = driver.FindElement(By.CssSelector("body")).Text;
            if (bodyText.Contains(text, StringComparison.Ordinal))
            {
                return true;
            }

            if (driver is IJavaScriptExecutor js)
            {
                js.ExecuteScript("window.scrollTo(0, document.body.scrollHeight);");
            }

            Thread.Sleep(300);
        }

        return false;
    }
}

File: tests/fixtures/selenium-chaos-suite/src/core/WaitUtils.ts
import { By, until, WebDriver, WebElement } from "selenium-webdriver";

export class WaitUtils {
  static async sleep(ms: number) {
    await new Promise((resolve) => setTimeout(resolve, ms));
  }

  static async waitVisible(driver: WebDriver, locator: By, timeout = 20000) {
    const el = await driver.wait(until.elementLocated(locator), timeout);
    await driver.wait(until.elementIsVisible(el), timeout);
    return el;
  }

  static async waitEnabled(driver: WebDriver, element: WebElement, timeout = 10000) {
    const started = Date.now();
    while (Date.now() - started < timeout) {
      try {
        const enabled = await element.isEnabled();
        if (enabled) return true;
      } catch (e) {
        // ignore to keep polling
      }
      await this.sleep(250);
    }
    return false;
  }

  static async bruteForceClick(driver: WebDriver, locator: By, tries = 5) {
    for (let i = 0; i < tries; i++) {
      try {
        const el = await driver.findElement(locator);
        await el.click();
        return;
      } catch (e) {
        await this.sleep(500);
      }
    }
  }

  static async scrollUntilTextAppears(driver: WebDriver, text: string, rounds = 12) {
    for (let i = 0; i < rounds; i++) {
      const bodyText = await driver.findElement(By.css("body")).getText();
      if (bodyText.includes(text)) return true;
      await driver.executeScript("window.scrollTo(0, document.body.scrollHeight)");
      await this.sleep(300);
    }
    return false;
  }
}
